dll/win32/shell32/CCopyAsPathMenu.cpp

```diff
@@ -0,0 +1,228 @@
+        {
+            CComPtr<IShellItem> item;
+            hr = array->GetItemAt(i, &item);
+            if (SUCCEEDED(hr))
+            {
+                CComHeapPtr<WCHAR> path;
+                hr = item->GetDisplayName(SIGDN_DESKTOPABSOLUTEPARSING, &path);
+                if (SUCCEEDED(hr))
+                {
+                    AppendToPathList(paths, path, i);
+                }
+            }
+        }
+    }
+#else
+    FIXME("Implement and use SHCreateShellItemArrayFromDataObject\n");
+    CDataObjectHIDA pCIDA(pdto);
+    HRESULT hr = pCIDA.hr();
+    if (SUCCEEDED(hr))
+    {
+        for (i = 0, count = pCIDA->cidl; i < count && SUCCEEDED(hr); ++i)
+        {
+            PCUIDLIST_ABSOLUTE folder = HIDA_GetPIDLFolder(pCIDA);
+            PCUIDLIST_RELATIVE item = HIDA_GetPIDLItem(pCIDA, i);
+            CComHeapPtr<ITEMIDLIST> full;
+            hr = SHILCombine(folder, item, &full);
+            if (SUCCEEDED(hr))
+            {
+                PCUITEMID_CHILD child;
+                CComPtr<IShellFolder> sf;
+                hr = SHBindToParent(full, IID_PPV_ARG(IShellFolder, &sf), &child);
+                if (SUCCEEDED(hr))
+                {
+                    STRRET strret;
+                    hr = sf->GetDisplayNameOf(child, SHGDN_FORPARSING, &strret);
+                    if (SUCCEEDED(hr))
+                    {
+                        CComHeapPtr<WCHAR> path;
+                        hr = StrRetToStrW(&strret, child, &path);
+                        if (SUCCEEDED(hr))
+                        {
+                            AppendToPathList(paths, path, i);
+                        }
+                    }
+                }
+            }
+        }
+    }
+    else
+    {
+        FORMATETC fmte = { CF_HDROP, NULL, DVASPECT_CONTENT, -1, TYMED_HGLOBAL };
+        STGMEDIUM stgm;
+        hr = pdto->GetData(&fmte, &stgm);
+        if (SUCCEEDED(hr))
+        {
+            for (i = 0, count = DragQueryFileW((HDROP)stgm.hGlobal, -1, NULL, 0); i < count && SUCCEEDED(hr); ++i)
+            {
+                WCHAR path[MAX_PATH];
+                if (DragQueryFileW((HDROP)stgm.hGlobal, i, path, _countof(path)))
+                {
+                    AppendToPathList(paths, path, i);
+                }
+            }
+            ReleaseStgMedium(&stgm);
+        }
+    }
+#endif
+
+    if (SUCCEEDED(hr))
+    {
+        DWORD err = SetClipboardFromString(paths);
+        hr = HRESULT_FROM_WIN32(err);
+    }
+
+    return hr;
+}
+
+STDMETHODIMP
+CCopyAsPathMenu::QueryContextMenu(HMENU hMenu, UINT indexMenu, UINT idCmdFirst, UINT idCmdLast, UINT uFlags)
+{
+    MENUITEMINFOW mii;
+    UINT Count = 0;
+
+    TRACE("CCopyAsPathMenu::QueryContextMenu(%p %p %u %u %u %u)\n", this,
+          hMenu, indexMenu, idCmdFirst, idCmdLast, uFlags);
+
+    // Detect SHIFT key press
+    // FIXME: GetKeyState shouldn't be called here, clients need to set the CMF_EXTENDEDVERBS flag.
```

Comment:
Is this CMF_... thing something we don't fully implement in our shell?

Reply:
I will fix CMF_EXTENDEDVERBS, no need to add this workaround here.

Reply:
`GetKeyState` workaround removed. I'll wait for your PR.